File: 104-maximum-depth-of-binary-tree/104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        def dfs(node: TreeNode | None, depth: int) -> int:
            if not node:
                return depth

            return max(dfs(node.left, depth + 1), dfs(node.right, depth + 1))
        
        return dfs(root, 0)
    
#        # BFS Solutions:   
#        if not root:
#            return 0
        
#         # BFS - nested list comprehension
#         level = [root]
#         depth = 0
        
#         while level:
#             level = [child for node in level for child in (node.left, node.right) if child is not None]
#             depth += 1
        
#         # BFS - queue
#         level = deque([root])
#         depth = 0
        
#         while level:
#             for _ in range(len(level)):
#                 node = level.popleft()
#                 if node.left:
#                     level.append(node.left)
#                 if node.right:
#                     level.append(node.right)
#             depth += 1
            
#         return depth

File: 104-maximum-depth-of-binary-tree/104-maximum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    // return dfs(root, 0)
    return bfs(root)
};

function dfs(node, depth) {
    if (!node) {
        return depth
    }
    return Math.max(dfs(node.left, depth + 1), dfs(node.right, depth + 1))
}

function bfs(node) {
    if (!node) {
        return 0
    }
    
    let depth = 0
    let pseudoQ = [node]
    
    while (pseudoQ.length) {
        const nextQ = []
        
        pseudoQ.forEach((node) => {
            node.left && nextQ.push(node.left)
            node.right && nextQ.push(node.right)
        })
        
        depth += 1
        pseudoQ = nextQ
    }
    
    return depth
}